assembly/index.ts

```diff
@@ -51,73 +77,95 @@ function load8(ptr: usize, offset: usize): u8 {
   return load<u8>(ptr + offset);
 }
 
-function hashBlocks(wPtr: usize, pPtr: usize, pos: u32, len: u32): u32 {
-  let
-    a: u32, b: u32, c: u32, d: u32,
-    e: u32, f: u32, g: u32, h: u32,
-    u: u32, i: u32, j: u32,
-    t1: u32, t2: u32,
-    k = K.dataStart;
-
-  while (len >= 64) {
-
-    a = H0;
-    b = H1;
-    c = H2;
-    d = H3;
-    e = H4;
-    f = H5;
-    g = H6;
-    h = H7;
-
-    for (i = 0; i < 16; i++) {
-      j = pos + i * 4;
-      store32(wPtr, i,
-        (<u32>load8(pPtr, j + 0) << 24) |
-        (<u32>load8(pPtr, j + 1) << 16) |
-        (<u32>load8(pPtr, j + 2) <<  8) |
-        (<u32>load8(pPtr, j + 3) <<  0)
-      );
-    }
 
-    for (i = 16; i < 64; i++) {
-      u  = load32(wPtr, i - 2);
-      t1 = rotr(u, 17) ^ rotr(u, 19) ^ (u >>> 10);
-      u  = load32(wPtr, i - 15);
-      t2 = rotr(u, 7) ^ rotr(u, 18) ^ (u >>> 3);
+@inline
+function CH(x: u32, y: u32, z: u32): u32 {
+  return((x & y) ^ (~x & z));
+}
 
-      store32(wPtr, i, t1 + load32(wPtr, i - 7) + t2 + load32(wPtr, i - 16));
-    }
+@inline
+function MAJ(x: u32, y: u32, z:u32): u32 {
+  return ((x & y) ^ (x & z) ^ (y & z));
+}
 
-    for (i = 0; i < 64; i++) {
-      t1 = (rotr(e, 6) ^ rotr(e, 11) ^ rotr(e, 25)) + ((e & f) ^ (~e & g)) + h + load32(k, i) + load32(wPtr, i);
-      t2 = (rotr(a, 2) ^ rotr(a, 13) ^ rotr(a, 22)) + ((a & b) ^ (a & c) ^ (b & c));
-      h = g;
-      g = f;
-      f = e;
-      e = d + t1;
-      d = c;
-      c = b;
-      b = a;
-      a = t1 + t2;
-    }
+@inline
+function EP0(x: u32): u32 {
+  return rotr(x, 2) ^ rotr(x, 13) ^ rotr(x, 22);
+}
 
-    H0 += a;
-    H1 += b;
-    H2 += c;
-    H3 += d;
-    H4 += e;
-    H5 += f;
-    H6 += g;
-    H7 += h;
-
-    pos += 64;
-    len -= 64;
-  }
+@inline
+function EP1(x: u32): u32 {
+  return rotr(x, 6) ^ rotr(x, 11) ^ rotr(x, 25);
+}
+
+@inline
+function SIG0(x: u32): u32 {
+  return rotr(x, 7) ^ rotr(x, 18) ^ (x >>> 3);
+}
 
-  return pos;
+@inline
+function SIG1(x: u32): u32 {
+  return rotr(x, 17) ^ rotr(x, 19) ^ (x >>> 10);
 }
 
+/**
+ * Expand message blocks (16 32bit blocks), into extended message blocks (64 32bit blocks),
+ * Apply SHA256 compression function on extended message blocks
+ * Update intermediate hash values
+ * @param wPtr pointer to expanded message block memory
+ * @param mPtr pointer to message block memory
+ */
+function hashBlocks(wPtr: usize, mPtr: usize): void {
+  a = H0;
+  b = H1;
+  c = H2;
+  d = H3;
+  e = H4;
+  f = H5;
+  g = H6;
+  h = H7;
+
+  // Load message blocks into first 16 expanded message blocks
+  for (i = 0; i < 16; i++) {
+    store32(wPtr, i,
+      load32Bswap(mPtr, i)
```

Comment:
how bout rename `load32Bswap` to `load32be` (big endian)?

Reply:
:+1: